docs/src/main/asciidoc/cdi-reference.adoc

```diff
@@ -661,6 +661,38 @@ This situation is very common when trying to use CDI with alternative JVM langua
 
 Quarkus however, can overcome these limitations when `quarkus.arc.transform-unproxyable-classes` is set to `true` (which is the default value).
 
+=== Container-managed Concurrency
+
+There is no standard concurrency control mechanism for CDI beans.
+Nevertheless, a bean instance can be shared and accessed concurrently from multiple threads.
+In that case it should be thread-safe.
+You can use standard Java constructs (`volatile`, `synchronized`, `ReadWriteLock`, etc.) or let the container control the concurrent access. 
+
+.Container-managed Concurrency Example
+[source,java]
+----
+import io.quarkus.arc.Lock;
+
+@Lock <1>
+@ApplicationScoped
+class SharedService {
+
+  void addAmount(BigDecimal amout) {
+    // ...changes some internal state of the bean
+  }
+  
+  @Lock(value = Lock.Type.READ, time = 1, unit = TimeUnit.SECONDS) <2> <3>
+  BigDecimal getAmount() {
+    // ...it is safe to read the value concurrently
+  }
+}
+----
+<1> `@Lock` (which maps to `@Lock(Lock.Type.WRITE)`) declared on the class instructs the container to lock the bean instance for any invocation of any business method, i.e. the client has "exclusive access" and no concurrent invocations will be allowed.
+<2> `@Lock(Lock.Type.READ)` overrides the value specified at class level. It means that any number of clients can invoke the method concurrently, unless the bean instance is locked by `@Lock(Lock.Type.WRITE)`. 
+<3> You can also specify the "wait time". If it's not possible to acquire the lock in the given time a `LockException` is thrown.
+
+TIP: `io.quarkus.arc.Lock` is a regular interceptor binding and as such can be used for any bean with any scope. However, it is especially useful for "shared" scopes, e.g. `@Singleton` and `@ApplicationScoped`.
```

Comment:
@mkouba, I don't have knowledge about CDI, so this might be a basic question - Assuming this `@Lock` is declared on a request scoped bean, do such beans have more than one (bean) instance at any given time? If yes, then does this locking mechanism apply per bean instance (i.e. a WRITE lock on bean instance1 would still allow a WRITE access on bean instance2 of the same bean type)? Or is the lock applicable for all instances of the same bean type?

Reply:
> If yes, then does this locking mechanism apply per bean instance (i.e. a WRITE lock on bean instance1 would still allow a WRITE access on bean instance2 of the same bean type)?

Yes, each bean instance has a separate lock.

Reply:
Do you think we should expand this part of the documentation to state that request scoped beans won't share the lock instances?

Reply:
Well, I can try to change the wording a bit. But `@Lock` is not a good match for `@RequestScoped` because there is exactly one bean instance per thread for `@RequestScoped`. In other words, you should not need locking there...

Reply:
Do endpoints (like this random example[1]) get marked (implicitly) as (request scoped) CDI beans? The use case that I am imagining, that might need clarity in the documentation, is something like this:
```
@Path("/hello")
public class HelloResource {

    @POST
    @Lock
    public String addCredit() {
        // some write operation
    }
}
```
and then the `@Lock` wouldn't work in such cases (for multiple simulatenous requests)?


[1] https://github.com/quarkusio/quarkus/blob/master/extensions/rest-client-jaxb/deployment/src/test/java/io/quarkus/restclient/jaxb/deployment/HelloResource.java

Reply:
I see. Well, JAX-RS resources are `@Singleton` by default in Quarkus. Of course, you can annotate them with `@RequestScoped` or set `quarkus.resteasy.singleton-resources=false`. In any case, I don't think it's a good idea to lock a _resource method_ like this.

Reply:
> Well, JAX-RS resources are @Singleton by default in Quarkus. 

Thank you for that detail.

> Of course, you can annotate them with @RequestScoped or set quarkus.resteasy.singleton-resources=false. In any case, I don't think it's a good idea to lock a resource method like this.

Given that the regular/common use case doesn't create an implicit `@RequestScoped`, as you explained, I think the current documentation is good enough. I believe people adding `@RequestScoped` will be knowing what that means semantically.

Thank you for answering the questions.

Reply:
You're welcome ;-)